### DISCUSSION TITLE
UX improvement - "smart" clips

### DISCUSSION THREAD

**Discussion by jounih** [audacity/audacity#1012] [Near future]
After more defined visual clip boundaries (rounded edges) and a top affordance for being able to drag and drop clips, it would also be useful to be able to adjust clip length non-destructively (like in many other DAWs).

<img width="1440" alt="smart clip 1@2x" src="https://user-images.githubusercontent.com/10109867/122547832-83985580-d028-11eb-83cc-51e0fbcfec1a.png">

<img width="1440" alt="smart clip 2@2x" src="https://user-images.githubusercontent.com/10109867/122547831-82ffbf00-d028-11eb-9599-324b99350ddb.png">

<img width="1440" alt="smart clip 3@2x" src="https://user-images.githubusercontent.com/10109867/122547792-754a3980-d028-11eb-9e8f-1444de9bfe89.png">

<img width="1440" alt="smart clip 4@2x" src="https://user-images.githubusercontent.com/10109867/122547829-8135fb80-d028-11eb-8f73-3fc1c4e97d1f.png">

Link to a prototype video:

https://vimeo.com/562162689#t=14m03s

Thoughts?

**Comment by DavidBailes** [comment]
There's a problem with the rounded corners. Assuming waveforms are displayed in a region which is bound by the bottom of the clip, then the bottom corners could cut of the displayed waveform. Or would the display region start just above the bottom of the clip?

**Reply by jounih** [reply]
Good point - though I think that would be a fairly uncommon circumstance. Incidentally the rounded bottom corners have been too difficult to implement with wxWidgets I've been told, so they are not rounded at the bottom for now, just at the top - I'll need to update the visuals.

**Reply by DavidBailes** [reply]
I agree that it would be uncommon - though might also occur when the waveform is zoomed in vertically. However, in the cases where it did occur, I think that being able to see the waveform should have priority over looking pleasing to the eye.

**Comment by DavidBailes** [comment]
Currently, if you drag from the start or end of a time selection, you drag the start/end of the time selection. If the start/end of a time selection coincides with the start/end of a clip how would the user indicate whether they wanted to grow/shrink the clip or change the time selection?

**Reply by jounih** [reply]
I can see there could be overlap - hmm. One fix I can think of is to have the clip trimming "handle" (which could be invisible) area slightly outside the clip, rather than inside. Or a separate tool mode (selection tool) for waveform selection - would like to avoid that if possible though

**Comment by SteveDaulton** [comment]
Perhaps the normal selection resize pointer when within the clip:
![resize-selection](https://user-images.githubusercontent.com/11681097/122571256-712c1500-d044-11eb-9084-5dd3f74497c3.png)

and a clip drag handle appear when the pointer is just outside the clip:
![drag-handle](https://user-images.githubusercontent.com/11681097/122571348-8acd5c80-d044-11eb-9a36-766d445bd035.png)

However, if the left end of the clip is at time = 0, then the drag handle would need to appear to the right of the edge, otherwise it will make it impossible to access the vertical ruler which is also a click target:
![drag-handle-left](https://user-images.githubusercontent.com/11681097/122571923-26f76380-d045-11eb-913b-c6ac4676a3d5.png)

**Comment by Paul-Licameli** [comment]
Remember the too little known thing I added to Audacity years ago:  whenever two things can hit test at the same mouse position, you can hit the esc key, and the cursor can change to the target that has the lower priority.

**Comment by Estecka** [comment]
Will there be a visual cue that a clip has been trimmed ? For example, something along the line of those arrows that appear at the edge when a clip begins before zero :
![Screenshot_1](https://user-images.githubusercontent.com/39855800/122607669-3a252600-d07b-11eb-8928-9ee8f8d6dc52.png)

**Reply by werame** [reply]
As far as I can tell there isn't one as implemented in 3.1.3, but I agree one is probably desirable.

**Reply by SteveDaulton** [reply]
> Will there be a visual cue that a clip has been trimmed ?

See: https://github.com/audacity/audacity/issues/2271

**Comment by werame** [comment]
'twas implemented in https://github.com/audacity/audacity/commit/c98222afc28ed9622f28dc190a7dae1770b4ac64 in case anybody still wonders. (So, it's in Audacity 3.1.) Doing a "mix and render to new track" will freeze the underlying wave as expected too, i..e. turns it into a destructive trim. 

Unlike in Adobe Audition though, there's not yet an easy ability in Audacity [slide](https://youtu.be/8WRKnhsX1bE?t=22) the "window" that the smart clip has into the underlying waveform, while keeping the window size constant. In Audacity, you have to adjust both ends instead.
